constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -1216,4 +1289,157 @@ Z: [10, 4294967050] & 0xffffffff;
             RangeConstraint::from_range(GoldilocksField::from(3), GoldilocksField::from(5))
         );
     }
+
+    #[test]
+    fn test_substitute_by_unknown_linear_to_quadratic() {
+        let mut expr = Qse::from_unknown_variable("x");
+        let y = Qse::from_unknown_variable("y");
+        let z = Qse::from_unknown_variable("z");
+
+        let three = Qse::from(GoldilocksField::from(3));
+        let subst = y.clone() * z.clone() + three.clone();
+        expr.substitute_by_unknown(&"x", &subst);
+
+        assert_eq!(expr.to_string(), "(y) * (z) + 3");
+    }
+
+    #[test]
+    fn test_substitute_by_unknown_inside_symbolic_expression() {
+        let mut expr = Qse::from_known_symbol("A", Default::default());
+        let subst = Qse::from_unknown_variable("x") * Qse::from_unknown_variable("y");
+
+        expr.substitute_by_unknown(&"A", &subst);
+
+        assert_eq!(expr.to_string(), "(x) * (y)");
+    }
+
+    #[test]
+    fn test_substitute_by_unknown_inside_quadratic_term() {
+        let x = Qse::from_unknown_variable("x");
+        let y = Qse::from_unknown_variable("y");
+        let mut expr = x.clone() * y.clone(); // x * y
+
+        let a = Qse::from_unknown_variable("a");
+        let one = Qse::from(GoldilocksField::from(1));
+        let subst = a.clone() + one; // x = a + 1
+
+        expr.substitute_by_unknown(&"x", &subst);
+
+        assert_eq!(expr.to_string(), "(a + 1) * (y)");
+    }
+
+    #[test]
+    fn test_substitute_by_unknown_quadratic_self_product() {
+        let x = Qse::from_unknown_variable("x");
+        let mut expr = x.clone() * x.clone(); // x * x
+
+        let a = Qse::from_unknown_variable("a");
+        let b = Qse::from_unknown_variable("b");
+        let subst = a.clone() + b.clone(); // x = a + b
+
+        expr.substitute_by_unknown(&"x", &subst);
+
+        assert_eq!(expr.to_string(), "(a + b) * (a + b)");
+    }
+
+    #[test]
+    fn test_substitute_linearly_then_result_becomes_quadratic() {
+        let mut expr = Qse {
+            quadratic: vec![],
+            linear: BTreeMap::from([("x", SymbolicExpression::from(GoldilocksField::from(3)))]),
+            constant: SymbolicExpression::from(GoldilocksField::from(0)),
+        };
+
+        let a = Qse::from_unknown_variable("a");
+        let b = Qse::from_unknown_variable("b");
+        let subst = a.clone() * b.clone(); // x = a * b
```

Comment:
```suggestion
        let subst = Qse::from_unknown_variable("a") * Qse::from_unknown_variable("b");
```

I think I'd also introduce a helper function in the test module to ergonomically build expressions, e.g. `3 * (x + y)` could look like:
`let expr = const(3) * (unknown("x") + unknown("y"))`

Just to reduce verbosity and keep the tests readable. See the [solver tests](https://github.com/powdr-labs/powdr/blob/252c646303cd65b1fd608bd1a496bb49fde30835/constraint-solver/tests/solver.rs#L68). Actually, the two modules should probably share a test utils module, but that can be future refactoring PR...

Reply:
Done.